### PR TITLE
Draft extended view for NDT7 using materialized joins, including Geo field renaming

### PR DESCRIPTION
**Pull request by mattmathis**
THIS PR IS NOT READY FOR FORMAL REVIEW.
It does not cover uploads at all.

Please reflect on the suitability of various things included here.

Primary changes:
- switch ndt7 to use gardner materialized joins
- change Geo structs column names from snake case to camel case and add missing fields
- deprecate test_date

Changes to facilitate user transition:
- Create unified_legcy_downloads which is (supposed to be) compatible with any valid query against the current unified_downloads.   I do not have a test for this assertion.
- Create _Geo containing the legacy field names.   Users can access the old fields by inserting underscore into their queries as needed.

---
This change is [Reviewable](https://reviewable.io/reviews/m-lab/etl-schema/89)

### CODE REVIEW

**Review by mattmathis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/m-lab/etl-schema/89)* status: 0 of 1 approvals obtained

___
*[views/library/referenced-by/ndt_unified_ndt5_downloads.sql, line 130 at r1](https://reviewable.io/reviews/m-lab/etl-schema/89#-MJisRo67TCXs28SXUJO:-MJisRo67TCXs28SXUJP:b-gfnbbr) ([raw file](https://github.com/m-lab/etl-schema/blob/06a93e106fec3662fe8d198e7f1e4a3c78757b31/views/library/referenced-by/ndt_unified_ndt5_downloads.sql#L130)):*
> ```TSQL
>         client.Network.Systems[SAFE_OFFSET(0)].ASNs[SAFE_OFFSET(0)] AS ASNumber,
>         '' AS ASName, -- MISSING
>         False AS Missing, -- MISSING
> ```

ASName and ASNumber should both be Null, but Nulls are treated as missing numbers, and break the UNION.  (4 places).
___
*[views/library/referenced-by/ndt_unified_ndt7_downloads.sql, line 17 at r1](https://reviewable.io/reviews/m-lab/etl-schema/89#-MJit75Y4OazzsIg20Pr:-MJit75Z6yWbxpgyDz_A:bf5zck5) ([raw file](https://github.com/m-lab/etl-schema/blob/06a93e106fec3662fe8d198e7f1e4a3c78757b31/views/library/referenced-by/ndt_unified_ndt7_downloads.sql#L17)):*
> ```TSQL
>   False AS IsErrored,
>   TIMESTAMP_DIFF(raw.Download.EndTime, raw.Download.StartTime, MICROSECOND) AS connection_duration
>  FROM   `mlab-oti.ndt.ndt7`
> ```

This should be mlab-oti.joined_ndt.ndt7

**Review by stephen-soltesz — COMMENTED** [review]
Regarding `unified_legacy_downloads`, it sounds like we would only have one prior version. Rather than "legacy" could we call this `unified_downloads_previous`? or `_prior`? or `_deprecated`?

Regarding `_Geo` or `new_Geo`, is this needed if `unified_legacy_*` is backward compatible? We don't need two ways to do the same thing here. Also, see my comment below about new names working now and in the future. I think we only want one Geo struct with new and old names.

> *[Reviewable](https://reviewable.io/reviews/m-lab/etl-schema/89#-:-MK7-HlH3T0hEVwpAQ8E:bszxcx)* status: 0 of 1 approvals obtained (waiting on @mattmathis and @stephen-soltesz)

___
*[views/library/referenced-by/ndt_unified_ndt5_downloads.sql, line 124 at r2](https://reviewable.io/reviews/m-lab/etl-schema/89#-MK72_dEEw0KE6uM2r7U:-MK72_dEEw0KE6uM2r7V:b-whsorq) ([raw file](https://github.com/m-lab/etl-schema/blob/95d69f903bb322a06fd84e34ce42e0804d37b893/views/library/referenced-by/ndt_unified_ndt5_downloads.sql#L124)):*
> ```TSQL
>              client.Geo.radius, -- aka AccuracyRadiusKm
>              FALSE AS Missing -- Future missing record flag
>       ) AS new_Geo,  -- Future primary Geo
> ```

Like the migration from `test_date` to `date`, we want the new name to work as it will in the future. I think this indicates adding all snake case fields to the `Geo` record, so that the camel case are added and the snake case remain and can be eliminated in the future. This may make the "unified_legacy*" unnecessary for the Geo fields.
___
*[views/library/referenced-by/ndt_unified_ndt5_downloads.sql, line 125 at r2](https://reviewable.io/reviews/m-lab/etl-schema/89#-MK73WdWCCEQBRS8V-pE:-MK73WdWCCEQBRS8V-pF:b-8v35uv) ([raw file](https://github.com/m-lab/etl-schema/blob/95d69f903bb322a06fd84e34ce42e0804d37b893/views/library/referenced-by/ndt_unified_ndt5_downloads.sql#L125)):*
> ```TSQL
>              FALSE AS Missing -- Future missing record flag
>       ) AS new_Geo,  -- Future primary Geo
> #      Client.Network -- BUG still old schema
> ```

What is the bug?

**Comment by mattmathis** [comment]
Yes, I am going to dated legacy views, with the date last.

With a few exceptions (e.g. test_date->date), I don't want users to use  mixed old and new names in the same query.  Creating dated legacy views gives users time to update their entire query to new columns, and is clearer that thy must do so.

File: views/intermediate_ndt/referenced-by/extended_ndt5_downloads.sql [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+      S2C.ClientPort AS Port,
+      Client.Geo,  -- Legacy Geo is first, to be removed
+      -- Legacy Geo approximates dev.maxmind.com/geoip/geoip2/geoip2-city-country-csv-databases/
+      STRUCT ( -- Future primary Geo
+             client.Geo.continent_code, -- aka ContinentCode
+             client.Geo.country_code, -- aka CountryCode
+             client.Geo.country_code3, -- aka CountryCode3
+             client.Geo.country_name, -- aka CountryName
+             client.Geo.region, -- aka Region
+             '' AS Subdivision1ISOCode, -- MISSING
```
Comment (critzo) [inline views/intermediate_ndt/referenced-by/extended_ndt5_downloads.sql:112]: This is actually the value currently in client.Geo.region

**Pull request closed without merging**